### PR TITLE
add draw_assignment, draw_sample, and expand draw_data functions

### LINKED ISSUE #347: draw_assignment and draw_sample
two functions similar to draw_estimates_data that 

1) draw_assignment: take data and then add assignment variable(s) (and assignment probs), running all assignment steps
2) draw_sample: take data and subset to sample (with inclusion probs), potentially more than once

### PR DESCRIPTION
**Pull request by graemeblair**
This PR closes https://github.com/DeclareDesign/DeclareDesign/issues/347

Adds:
- `draw_sample` which takes data and samples either all sampling steps (i.e. it subsets) by default, or just some of them using start and end step arguments;
- `draw_assignment` which takes data and assigns either all assignment steps (by default), or just some of them based on start and end step arguments;
- `draw_data` which takes data and conducts further data steps (either all of them by default, or a subset of them). Note that a design can technically have just a bunch of sampling and assignment steps and need not have a population or data at the top (given the way + works; we tested this), so it does make sense for you to be able to go from 1 to the end of the design starting with data you give. **`draw_data` by default runs the whole design, so now does or does not take data.**

Moved files and documentation around to more clearly document the `get_*` and `draw_*` steps.

### CODE REVIEW

**Comment by graemeblair** [comment]
@macartan do you mind taking a look?

File: R/get_functions.R [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  
+  if(end < 1 || end > length(design)){
+    stop("Please provide an end step as a number between 1 and the total number of steps in the design.")
+  }
+  
+  design_subset <- Filter(function(x) attr(x, "causal_type") == "dgp", design[start:end])
+  
+  run_design_internal.design(design_subset, current_df = data, results = list(current_df = 0))$current_df
+  
+}
```
Comment (nfultz) [inline R/get_functions.R]: These are all very similar, pull out the common logic to a helper function.

```
get_data_imp <- function(design, data = NULL, start = 1, end = length(design), pred) {

   if(is.null(data)){
    stop("Please provide a data frame to the data argument.")
  }

   if(start < 1 || start > length(design)){
    stop("Please provide a starting step as a number between 1 and the total number of steps in the design.")
  }

   if(end < 1 || end > length(design)){
    stop("Please provide an end step as a number between 1 and the total number of steps in the design.")
  }

   design_subset <- Filter(pred, design[start:end])

   run_design_internal.design(design_subset, current_df = data, results = list(current_df = 0))$current_df

 }

get_data <- function(design, data = NULL, start = 1, end = length(design)) {
  get_data_impl(design, data = NULL, start = 1, end = length(design),
    function(x) attr(x, "causal_type") == "dgp")
}
```

**Review by nfultz — COMMENTED** [review]
Need to tidy up some of the duplicated logic.

File: R/get_functions.R (hunk 1 above)
Comment (nfultz) [reply R/get_functions.R]: For the predicates, using `%in%` instead of `==` is more defensive.

File: R/construct_design.R [hunk 2]
```diff
@@ -21,6 +21,8 @@
 #'
 #' @return a list of two functions, the \code{design_function} and the \code{data_function}. The \code{design_function} runs the design once, i.e. draws the data and calculates any estimates and estimands defined in \code{...}, returned separately as two \code{data.frame}'s. The \code{data_function} runs the design once also, but only returns the final data.
 #'
+#' @name declare_design
```
Comment (nfultz) [inline R/construct_design.R:24]: Is declare_design come back from the grave?

File: R/design_print_summary.R [hunk 3]
```diff
@@ -0,0 +1,264 @@
+
```
Comment (nfultz) [inline R/design_print_summary.R:1]: I usually try to keep reorganization to a seperate PR, otherwise the diffs get large and hard to follow, and history breaks.

File: R/construct_design.R (hunk 2 above)
Comment (graemeblair) [reply R/construct_design.R:24]: Hah, no just naming the file :)

File: R/design_print_summary.R (hunk 3 above)
Comment (graemeblair) [reply R/design_print_summary.R:1]: ah, thanks, tried to avoid and kept in a separate PR. will do moving forward.

**Comment by graemeblair** [comment]
I think I implemented all this, but not understanding the sentinal value suggestion. 

`data %in% -9` works when you send -9 but when you send an actual data.frame it doesn't work given length:

```
In if (data %in% -9) { :
  the condition has length > 1 and only the first element will be used
```

Worry about false positives?

**Comment by nfultz** [comment]
Whoops, should be `data ==` there

Edit:

or rather `identical()`

**Comment by graemeblair** [comment]
ah got it thanks!
